aten/src/ATen/core/boxing/KernelFunction.h

```diff
@@ -47,9 +47,25 @@ TORCH_API void ambiguous_autogradother_kernel(OperatorKernel*, const OperatorHan
 class TORCH_API KernelFunction final {
 public:
   // This is how boxed kernels are actually stored
-  using InternalBoxedKernelFunction = void(OperatorKernel*, const OperatorHandle&, Stack*);
+  //
+  // Note [Plumbing Keys Through The Dispatcher]
+  // Benchmarks have shown that it is expensive for the dispatcher to read from thread-local storage (TLS)
+  // upon every dispatch call into order to compute which kernel to dispatch to.
+  //
+  // To mitigate this, we allow kernels to optionally take in the DispatchKeySet that the dispatcher computed,
+  // manually calculate the next kernel that should be dispatched to, and pass that information back to the dispatcher,
+  // thus bypassing the dispatcher overhead of reading from TLS to recompute keys.
+  // Kernels that opt into this convention must register themselves with Library::impl_withKeys() rather than Library::impl().
+  // See the kernels in VariableTypeEverything.cpp and TraceTypeEverything.cpp for examples.
+  //
+  // This changes the calling convention for the dispatcher. The dispatcher now expects all kernels that it stores
+  // to have a first argument of type DispatchKeySet. The mechanism for optionally passing that DispatchKeySet
+  // into the kernel lives in make_boxed_from_unboxed_functor.h.
+  // See Note [Plumbing Keys Through The Dispatcher 2] for details.
+  using InternalBoxedKernelFunction = void(OperatorKernel*, const OperatorHandle&, DispatchKeySet, Stack*);
```

Comment:
Good

Reply:
One high level comment about the comment above: the explanation here is slightly backwards. You first say "we allow kernels to optionally take in the DispatchKeySet that the dispatcher computed" and then later say "The dispatcher now expects all kernels that it stores to have a first argument of type DispatchKeySet". The latter information is the most important up-front information, since it explains what this type is! Then, as additional information, one can explain that in the user-visible API you can omit the DispatchKeySet from the kernel. But here, in the comment about `InternalBoxedKernelFunction`, the explanation of the type itself should come first.

Reply:
One thing that I can't easily tell from this comment is what exactly the invariants on DispatchKeySet are. You say this set is "what the dispatcher computed"; but what exactly is this set? Does it include TLS keys? (Yes) Is the dispatch key that we are currently dispatching to masked out? (No?) Is it guaranteed that all keys to the left of the current dispatch key are unset? (Yes?)

Operationally, the implementation is very simple, but it falls upon us to explain how to interpret the set as given here.